File: product-processor-service/src/main/java/com/example/product_processor_service/configuration/kafka/factory/producer/KafkaProducerFactoryConfiguration.java
package com.example.product_processor_service.configuration.kafka.factory.producer;

import com.example.product_processor_service.configuration.kafka.environment.KafkaEnvironment;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.ProducerFactory;

import java.util.HashMap;
import java.util.Map;

@Configuration
public class KafkaProducerFactoryConfiguration {
    private final KafkaEnvironment kafkaEnvironment;

    public KafkaProducerFactoryConfiguration(KafkaEnvironment kafkaEnvironment) {
        this.kafkaEnvironment = kafkaEnvironment;
    }

    @Bean
    public ProducerFactory<String, String> productUrlProducerFactory() {
        DefaultKafkaProducerFactory<String, String> cfg = new DefaultKafkaProducerFactory<>(buildAtLeastOnceProducerProperties());
        cfg.setKeySerializer(new StringSerializer());
        cfg.setValueSerializer(new StringSerializer());
        return cfg;
    }

    private Map<String, Object> buildAtLeastOnceProducerProperties() {
        Map<String, Object> producerProperties = new HashMap<>();
        producerProperties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaEnvironment.getKAFKA_BOOTSTRAP_SERVERS());
        producerProperties.put(ProducerConfig.ACKS_CONFIG, "1");
        producerProperties.put(ProducerConfig.RETRIES_CONFIG, kafkaEnvironment.getKAFKA_PRODUCER_RETRY());

        return producerProperties;
    }
}

File: product-manager-service/src/main/kotlin/com/example/product_manager_service/configuration/kafka/producer/factory/KafkaProducerFactoryConfiguration.kt
package com.example.product_manager_service.configuration.kafka.producer.factory

import com.example.product_manager_service.configuration.kafka.environment.KafkaEnvironment
import com.example.product_manager_service.model.product.dto.ProductPublisherDto
import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.ProducerFactory
import org.springframework.kafka.support.serializer.JsonSerializer


@Configuration
class KafkaProducerFactoryConfiguration(private val kafkaEnvironment: KafkaEnvironment) {
    @Bean
    fun productPublisherProducerFactory(objectMapper: ObjectMapper): ProducerFactory<String, ProductPublisherDto> {
        val cfg = DefaultKafkaProducerFactory<String, ProductPublisherDto>(buildAtLeastOnceProducerProperties())
        cfg.valueSerializer = JsonSerializer(objectMapper)
        return cfg
    }

    @Bean
    fun objectMapper(): ObjectMapper = ObjectMapper()

    private fun buildAtLeastOnceProducerProperties(): Map<String, *> {
        val producerProperties: MutableMap<String, Any> = HashMap()
        producerProperties.put(
            ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
            StringSerializer().javaClass.name
        )
        producerProperties.put(
            ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
            JsonSerializer::class
        )
        producerProperties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaEnvironment.KAFKA_BOOTSTRAP_SERVERS)
        producerProperties.put(ProducerConfig.ACKS_CONFIG, "1")
        producerProperties.put(ProducerConfig.RETRIES_CONFIG, kafkaEnvironment.KAFKA_PRODUCER_RETRIES)

        return producerProperties
    }
}